circle_ci/create_changelog.js

```diff
@@ -0,0 +1,128 @@
+ */
+const axios = require('axios')
+const childProcess = require('child_process')
+const fs = require('fs-promise')
+
+// ----------------------------------------
+// Environment
+// ----------------------------------------
+
+const { CIRCLE_BRANCH, CIRCLE_PROJECT_USERNAME, CIRCLE_PROJECT_REPONAME } = process.env
+if (CIRCLE_BRANCH) throw new Error('Missing env var CIRCLE_BRANCH.')
+if (CIRCLE_PROJECT_USERNAME) throw new Error('Missing env var CIRCLE_PROJECT_USERNAME.')
+if (CIRCLE_PROJECT_REPONAME) throw new Error('Missing env var CIRCLE_PROJECT_REPONAME.')
+
+// ----------------------------------------
+// Arg parse
+// ----------------------------------------
+const argv = require('yargs')
+  .usage('USAGE: $0 [options]')
+  .option('t', {
+    alias: 'token',
+    describe: 'github personal access token'
+  })
+  .help()
+  .argv
+
+// ----------------------------------------
+// Utils
+// ----------------------------------------
+const log = (...msgs) => console.log('CREATE CHANGELOG:', ...msgs) // eslint-disable-line no-console
+
+const wait = ms => new Promise(res => setTimeout(res, ms))
+
+// format axios responses for abbreviated logging
+const formatRes = (res) => {
+  if ({}.toString.call(res) !== '[object Object]') return
+  const { status, data } = res
+
+  let truncatedData = ''
+  if (data) truncatedData = data.length > 100 ? `${res.data.substr(0, 100)}...` : data
+
+  return JSON.stringify({ status, data: `${truncatedData}` }, null, 2)
+}
+
+// ----------------------------------------
+// Methods
+// ----------------------------------------
+let endpoint = `https://github-changelog-api.herokuapp.com/${CIRCLE_PROJECT_USERNAME}/${CIRCLE_PROJECT_REPONAME}`
+if (argv.token) endpoint += `?token=${argv.token}`
+
+const postChangelogRequest = () => {
+  log(`Posting changelog request: ${endpoint}`)
+  return axios.post(endpoint)
+    .catch(err => {
+      log('Failed to POST changelog request')
+      console.error(err)
+      process.exit(1)
+    })
+}
+
+const pollForChangelog = () => {
+  log('Polling for completed changelog')
+  return axios.get(endpoint)
+    .then(
+      res => {
+        // changelog is done
+        log('Received changelog string', formatRes(res))
+        return res.data
+      },
+      err => {
+        if (err.response.status === 302) {
+          // poll
+          log('Changelog job in progress, waiting...', formatRes(err.response))
+          return wait(5000).then(pollForChangelog)
+        } else {
+          // reject all other errors
+          log('Error while polling', formatRes(err.response))
+          process.exit(1)
+        }
+      }
+    )
+}
+
+const writeChangelog = (data) => {
+  log('Writing changelog to disk:', data)
+  return fs.writeFile('CHANGELOG.md', data)
+    .catch(err => {
+      log('Failed to write changelog to disk:', formatRes(err.response))
+      process.exit(1)
+    })
+}
+
+const commitAndPush = () => new Promise((resolve, reject) => {
```

Comment:
Hm, I don't think it should _always_ resolve as it then can't be moved around and relied on to resolve on success / reject on error.